core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java

```diff
@@ -567,6 +568,52 @@ default EventExecutor defaultSubscriberExecutor() {
         return FuseableStreamMessage.error(this, function);
     }
 
+    /**
+     * Peeks values emitted by this {@link StreamMessage} and applies the specified {@link Consumer}.
+     *
+     * <p>For example:<pre>{@code
+     * StreamMessage<Integer> source = StreamMessage.of(1, 2, 3, 4, 5);
+     * StreamMessage<Integer> ifEvenExistsThenThrow = source.peek(x -> {
+     *      if (x % 2 == 0) {
+     *          throw new IllegalArgumentException();
+     *      }
+     * });
+     * }</pre>
+     */
+    default StreamMessage<T> peek(Consumer<? super T> action) {
+        requireNonNull(action, "action");
+        final Function<T, T> function = obj -> {
+            action.accept(obj);
+            return obj;
+        };
+        return map(function);
+    }
+
+    /**
+     * Peeks values emitted by this {@link StreamMessage} and applies the specified {@link Consumer}.
+     *
+     * <p>For example:<pre>{@code
+     * StreamMessage<Number> source = StreamMessage.of(0.1, 1, 0.2, 2);
+     * StreamMessage<Number> ifEvenExistsThenThrow = source.peek(x -> {
+     *      if (x % 2 == 0) {
+     *          throw new IllegalArgumentException();
+     *      }
+     * }, Integer.class);
+     * }</pre>
+     */
+    default <U extends T> StreamMessage<T> peek(Consumer<? super U> action, Class<? extends U> type) {
```

Comment:
Q) Can you add additional javadoc on why this method is needed? (what this method does that is different from `StreamMessage#peek(Consumer..)`?

This is also for my reference since I don't understand this API fully as well (I find it a little surprising peek selectively peeks depending on type)
e.g. the following test fails
```
@Test
void peekWithType() {
    final StreamMessage<Number> source = StreamMessage.of(0.1, 1, 0.2, 2);
    final List<Number> collected = new ArrayList<>();
    final Consumer<Integer> ifEvenExistsThenThrow = collected::add;
    final StreamMessage<Number> peeked = source.peek(ifEvenExistsThenThrow, Integer.class);
    StepVerifier.create(peeked)
                        .expectNext(0.1, 1, 0.2, 2)
                        .verifyComplete();
    await().untilAsserted(() -> assertThat(collected).hasSize(4));
}
```

Reply:
```java
@Test
void peekWithType() {
    final StreamMessage<Number> source = StreamMessage.of(0.1, 1, 0.2, 2);
    final List<Number> collected = new ArrayList<>();
    final Consumer<Integer> ifIntegerExistsThenAdd = collected::add;

    // only 1, 2 (Integer) will be added to collected
    final StreamMessage<Number> peeked = source.peek(ifIntegerExistsThenAdd, Integer.class);

    StepVerifier.create(peeked)
                .expectNext(0.1, 1, 0.2 , 2)
                .verifyComplete();
    await().untilAsserted(() -> assertThat(collected).hasSize(2)); // collected size should be 2 (collected = [1, 2])
}
```
Based on your test, I think `colloected.size()` should be `2`!

I think your test looks better to understand!

If you understand why this `peek(action, type)` method is needed, then I'll update javadoc and this unit-test based on your test 😄

Reply:
(FYI) why peekWithType method is needed? See this [issue](https://github.com/line/armeria/issues/3097#issue-714255549)

Reply:
> (FYI) why peekWithType method is needed? See this issue cc. @jrhee17

Should've read the issue more carefully 😅  While I wasn't able to find any references that does this, it seems like a useful feature. Sounds good to me! 👍